### PR TITLE
Add drive-format script

### PR DESCRIPTION
**Pull request by LukeShortCloud**

### CODE REVIEW

**Comment by LukeShortCloud** [comment]
Intended use-case for this script:

- An administrator or an app can run `--check` to see if there is any data on the drive.
- If there is data, the app could prompt the user to let them know that there is data and confirm if they really want to erase all data.
- Then the script can be run with `--format` via the app if the end-user agrees to proceed if there is data found or automatically if there is no data found.

---

Example output of `--check`:

- Scenario 1 - Fedora OS drive

```
$ sudo bash /mnt/drive-format.sh --check /dev/vda
WARNING: Partition table does exist!
WARNING: Partition table is not empty!
WARNING: Partition(s) are mounted!
INFO: Partition named mount not found in /etc/fstab.
WARNING: Partition UUID mount found in /etc/fstab!
WARNING: Partition UUID mount found in /etc/fstab!
$ echo $?
1
```

- Scenario 2 - Flash drive with one partition on it

```
$ sudo bash /mnt/drive-format.sh --check /dev/vdb
WARNING: Partition table does exist!
WARNING: Partition table is not empty!
INFO: Partition(s) are not mounted.
$ echo $?
1
```

- Scenario 3 - New unformatted drive

```
$ sudo bash /mnt/drive-format.sh --check /dev/vdc
INFO: Partition table does not exist.
$ echo $?
0
```

Example output of `--format`:

- Scenario 1 - Drive with no active mounts

```
$ sudo bash /mnt/drive-format.sh --check /dev/vdb
INFO: Clearing the existing partition table and first partition metadata.
INFO: Creating a new GPT partition table.
INFO: Creating a partition using all available space.
INFO: Formatting the partition to ext4 with case folding support.
$ lsblk -o name,fstype,size | grep -P "NAME|vdb"
NAME   FSTYPE   SIZE
vdb               2G
└─vdb1 ext4       2G
```

- Scenario 2 - Drive with active mounts (formatting works but requires a reboot)

```
$ sudo bash /mnt/drive-format.sh --format /dev/vdb
INFO: Clearing the existing partition table and first partition metadata.
INFO: Creating a new GPT partition table.
Error: Partition(s) 1 on /dev/vdb have been written, but we have been unable to inform the kernel of the change, probably because it/they are in use.  As a result, the old partition(s) will remain in use.  You should reboot now before making further changes.
INFO: Creating a partition using all available space.
INFO: Formatting the partition to ext4 with case folding support.
```

---

Having slept on it, here are my thoughts on other improvements I can make:
- `--check`
    - Return code of "0" should be used if a partition table is found but empty. That is safe to use.
- `--format`
    - Error out if anything fails for the format.
    - `umount` any active mounts before formatting. This allows the file system to be used immediately instead of needing a reboot.
    - Test on microSD card and/or NVMe drive. I have only been testing on virtual and physical serial drives. The naming convention for partitions on microSD cards and NVMe drives is a bit different but the script was written in a way that should account for that (just need to verify it works at expected).

**Comment by alkazar** [comment]
IMO there is no need for `--check`. We should always prompt/warn the user about potential data destruction.
There is too much at stake if there is any kind of error in the logic.

File: sbin/drive-format.sh [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+if [[ "${1}" == "--check" ]]
+    # Use return code of "0" if it is safe to format or "1" if potentially unsafe.
+    then drive_format_check_rc=0
+    # The partition type reported by `parted` is one of three possible values: "gpt", "msdos" (MBR), or "unknown" (none).
+    if parted ${2} print 2> /dev/null | grep -q "Partition Table: unknown"
+        then echo "INFO: Partition table does not exist."
+    else
+        echo "WARNING: Partition table does exist!"
+        drive_format_check_rc=1
```
Comment (alkazar) [inline sbin/drive-format.sh:11]: Aren't we done at this point? `drive_format_check_rc` will be 1 no matter what, so no need for all the complexity below?

No one is intended to use this script directly, so all the logic to provide detailed information seems unnecessary.

**Comment by LukeShortCloud** [comment]
The original reasoning behind `--check` is to see if (1) the partition table exists and (2) if there are any partitions. For example, a new flash drive could be blank. If not, then we can assume there is no data at all on the device and move onto formatting. The only edge case I can think of is if an advanced Linux user has hidden partitions that are purposely not in the partition table.

The verbose logging was simply to help gather as much data as possible about the drive. It is there for logging purposes, not end-users.

I will go ahead and simply all this.

**Comment by LukeShortCloud** [comment]
I made some changes:
- Remove `--check`.
- `umount` drive partitions before formatting.
- Error out if the drive format fails.
- Add `partprobe` to ensure new partitions are loaded.

Tested and working on:
- Fedora virtual machine with a virtual serial drive
- Arch Linux on LCD Steam Deck with a microSD card

**Comment by LukeShortCloud** [comment]
The latest force push was to (1) update the `README.md` based on the simplification and (2) remove a useless code comment.

File: sbin/drive-format.sh [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        then echo "ERROR: Failed to format ${partition_to_format} to ext4!"
+        exit 1
+    else
+        echo "INFO: Successfully formatted ${partition_to_format} to ext4."
+    fi
+    echo "INFO: Reloading system partition information."
+    partprobe
+else
+    echo -e "Usage:
+\tdrive-format.sh --format /dev/\${DEVICE}\tForce the device to be formatted even if in-use"
```
Comment (alkazar) [inline sbin/drive-format.sh:29]: Is `--format` still used?

Comment (LukeShortCloud) [reply sbin/drive-format.sh:29]: No, not any more. I will update that. Thanks!

**Comment by LukeShortCloud** [comment]
We're moving this to `hwctl` which will be included in this repository soon.

**Comment by LukeShortCloud** [comment]
Related: https://github.com/playtron-os/playtron-os-scripts/pull/7

**Pull request closed without merging**